plugin/trino-snowflake/src/main/java/io/trino/plugin/snowflake/SnowflakeClient.java

```diff
@@ -0,0 +1,565 @@
+    }
+
+    private static SliceReadFunction variantReadFunction()
+    {
+        return (resultSet, columnIndex) -> Slices.utf8Slice(resultSet.getString(columnIndex).replaceAll("^\"|\"$", ""));
+    }
+
+    private static ColumnMapping columnMappingPushdown(ColumnMapping mapping)
+    {
+        if (mapping.getPredicatePushdownController() == PredicatePushdownController.DISABLE_PUSHDOWN) {
+            log.debug("SnowflakeClient.columnMappingPushdown: NOT_SUPPORTED mapping.getPredicatePushdownController() is DISABLE_PUSHDOWN. Type was " + mapping.getType());
+            throw new TrinoException(NOT_SUPPORTED, "mapping.getPredicatePushdownController() is DISABLE_PUSHDOWN. Type was " + mapping.getType());
+        }
+
+        return new ColumnMapping(mapping.getType(), mapping.getReadFunction(), mapping.getWriteFunction(), PredicatePushdownController.FULL_PUSHDOWN);
+    }
+
+    private static ColumnMapping timeColumnMapping(JdbcTypeHandle typeHandle)
+    {
+        int precision = typeHandle.getRequiredDecimalDigits();
+        checkArgument(precision <= SNOWFLAKE_MAX_SUPPORTED_TIMESTAMP_PRECISION, "The max timestamp precision in Snowflake is " + SNOWFLAKE_MAX_SUPPORTED_TIMESTAMP_PRECISION);
+        return ColumnMapping.longMapping(
+                TimeType.createTimeType(precision),
+                (resultSet, columnIndex) -> {
+                    LocalTime time = SNOWFLAKE_TIME_FORMATTER.parse(resultSet.getString(columnIndex), LocalTime::from);
+                    return Timestamps.round(time.toNanoOfDay() * PICOSECONDS_PER_NANOSECOND, 12 - precision);
+                },
+                timeWriteFunction(precision),
+                PredicatePushdownController.FULL_PUSHDOWN);
+    }
+
+    private static LongWriteFunction snowFlaketimeWriter(Type type)
+    {
+        return timeWriteFunction(((TimeType) type).getPrecision());
+    }
+
+    private static LongWriteFunction timeWriteFunction(int precision)
+    {
+        checkArgument(precision <= SNOWFLAKE_MAX_SUPPORTED_TIMESTAMP_PRECISION, "Unsupported precision: %s", precision);
+        String bindExpression = format("CAST(? AS time(%s))", precision);
+        return new LongWriteFunction()
+        {
+            @Override
+            public String getBindExpression()
+            {
+                return bindExpression;
+            }
+
+            @Override
+            public void set(PreparedStatement statement, int index, long picosOfDay)
+                    throws SQLException
+            {
+                picosOfDay = Timestamps.round(picosOfDay, 12 - precision);
+                if (picosOfDay == Timestamps.PICOSECONDS_PER_DAY) {
+                    picosOfDay = 0;
+                }
+                LocalTime localTime = LocalTime.ofNanoOfDay(picosOfDay / PICOSECONDS_PER_NANOSECOND);
+                // statement.setObject(.., localTime) would yield incorrect end result for 23:59:59.999000
```

Comment:
probably becuase the division can round the value.

`StandardColumnMappings#fromTrinoTime` handles this correctly and can be used instead of this separate logic.